MovingBlocks/Terasology
# Fix parallax and normal mapping

### Contains

This PR:
 1. Corrects the formulae used in the shaders to calculate parallax and normal maps.
 1. Restricts parallax maps to not fetch colours from beyond the current texture.
 1. Alters the default setting for parallax maps to improve their appearance.
 1. Enables the use of parallax mapping without normal mapping.

Parts 2 and 3 resolve #2418, and part 4 deals with some of the confusingness of the settings discussed in that issue.

The branch 4Denthusiast/Terasology:parallax uses a more realistic version but looks worse due to parallax mapping being unable to extend the block's rendering beyond its original size. This branch is a compromise which makes the block appear flatter at shallow angles.

### How to test

Activate parallax mapping in the video menu. Activating extra lighting as well should no longer be necessary.

Place some core:depthTest or core:ironBar blocks, and look at them from various angles.

To see that the parallax calculations are now more correct, it's easier to compare a094555 to develop, rather than the final version. In that version, it looks as though every block is a little window through onto the texture atlas. (This is more apparent in worlds with large-ish texture atlases i.e. many blocks. I used coreSampleGameplay.) This behaviour is correct (except the issues fixed in later commits), as it corresponds to realistic parallax. The current version in develop, by contrast, changes its appearance depending on camera angle, which is clearly incorrect.

### Outstanding before merging

- [ ] The effect of normal maps are more subtle, so I haven't managed to test that they're actually working properly.
- [ ] I'm not sure if it's appropriate to leave gaps of unused texture indices, or whether it would be better to reduce the index of textureAtlasHeight if textureAtlasNormal isn't used (e.g. WorldRelfectionNode line 179).
- [ ] I'd like some more opinions on whether 466a1d7 or 034b678 (realistic but more distorted, or compromise) is better.
- [ ] Is there any way to get linear interpolation for textures that's actually linear everywhere, rather than flat near the defined points? Setting textureAtlasHeight to use that would be better if possible.
